UAS CRM 19 Implemented crud user api and user auth

momopoi443 opened the pull request:
Implemented all layers of authentication part along with main part that manipulates Users.

Commit:
Implemented AuthController

Commit:
Implemented auth-specific Payloads

Commit:
Described JwtService and UserService contracts;
occasionally corrected code of AuthController, NotFoundException, Login TokenPair UserCreation - payloads;
Added security models

Commit:
Implemented JwtService
Changed for optimization purposes JwtService contract
For implementation added needed properties and jjwt dependency

Commit:
Refactored code to cleaner
Separated Auth logic into service
Added Redis to store refresh tokens
Added TODOs related to Andriy

Commit:
Polished code:
reorganised packages
accordingly changed pom
added spring security dependency
added logging
added exception handling
corrected small moments

Commit:
deleted properties for testing

Commit:
Refactored code of AuthServiceImpl and JwtServiceImpl
Rollback changes to user entity

Commit:
Added permissions and subject claims to AccessTokenClaims
Specified claims of RefreshTokenClaims
Added adapter for RefreshTokenClaims
Added package token instead of model to improve clarity

Commit:
Fixed error with Class<List<Permission>> casting in AccessTokenHandlerAdapter
Added TimeUtils with Date, LocalDateTime conversion methods

Commit:
Changed logout endpoint to accept path variable of uuid type
Enabled methods param validation in AuthController
Changed logout method of AuthService to accept user uuid
Changed refresh method of AuthService to corresponding changes in refresh token claims
Changed generation, parsing and invalidation(refresh) of access, refresh tokens to corresponding changes in their claims
Added getPermissionsList method to User entity class
Added findByUuid method to UserService interface

Commit:
Added @NotBlank validation constrains to payloads
Added TODO for handling validation exceptions

Commit:
AuthConfig: configured RedisTemplate using key as UUID and value String
Renamed UserCreationRequestPayload to UserRegistrationRequestPayload
Removed password coding from AuthServiceImpl register method
Optimized AccessTokenHandlerAdapter imports
Added TODO not to forget implement password coding in UserService

Commit:
refactored pom.xml:
- added more verbose description
- logically separated dependencies

Commit:
rewrite User entity along with liquibase changelogs

Commit:
Refactored JwtService and JwtServiceImpl:
- added and implemented issueTokenPair method
- added JwtProperties as folder of auth.jwt.* properties
- refactored JwtServiceImpl to use JwtProperties

Commit:
moved Postgres, Redis and JWT properties to env variables and returned server.port to random port

Commit:
Renamed AuthConfig to RedisConfig

Commit:
Moved prefix of AuthController path to @RequestMapping

Commit:
Refactored auth-service controller payloads:
- deleted TokenPairResponsePayload
- moved RegisterResponsePayload to user-manager controller payloads and removed validation
- moved UserRegistrationRequestPayload to user-manager controller payloads

Commit:
Moved FailedRegistrationException

Commit:
changed package of UserManagerApplicationTests to userauthservice

Commit:
changed package of ApiErrorInfo to userauthservice

Commit:
Moved token package from auth-service/service to auth-service/

Commit:
Correct name of service package in user-manager part
Described and implemented UserService:
- Implemented and added UserMapper
- Added UserRepository
- Added UserDTO
- Added InvalidArgumentException
- Implemented UserServiceImpl

Commit:
Implemented authorization:
- Described rules in AuthzConfig
- Implemented JwtAuthenticationFilter

Commit:
Implemented UserController

Commit:
Corrected package name of auth-service
Refactored AuthService and AuthServiceImpl:
- removed register method
- added password coding when checking in login method

Commit:
Added property for hibernate to disable ddl
Corrected name of join column in User entity

Commit:
sync commit

Commit:
Refactored User entity:
- now user have one role
- changed accordingly liquibase changelogs

Commit:
refactored users table name fields to snake case
changed service description in pom.xml

Commit:
Refactored Auth-service:
- removed path versioning from auth url
- renamed LoginRequestPayload to CredentialsDTO
- rewrote AuthServiceImpl to use UserRepository instead of UserService
- refactored JwtServiceImpl to take in User instead of UserDTO, changed exception messages
- changed AccessToken to have 1 role

Commit:
Refactored user-manager service:
- authz part would be rewritten to new requirements
- removed path versioning from UserController url
- refactored all payloads
- removed FailedRegistrationException
- added methods for mapping user from DTOs
- removed unused methods from UserRepository
- refactored UserService and UserServiceImpl accordingly for latest requirements
Added global validation exceptions handler

Commit:
Prepared for branch for pushing:
Removed unrelated authz part

Commit:
Fixed bugs

Commit:
Added readableName static method to all exception and accordingly changed controller advices
Synchronized changed with entity implementation branch
Wrote tests for User- controller, mapper and service

Commit:
Small test fix

zhenyaAreYouHere63 commented on user-manager-service/src/main/java/com/crmfoodestablishment/usermanager/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,64 @@
+import org.hibernate.validator.constraints.UUID;
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
```
Don't use wildcard imports

zhenyaAreYouHere63 commented on user-manager-service/src/main/java/com/crmfoodestablishment/usermanager/crud/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
```
here too wildcard import

zhenyaAreYouHere63 commented on user-manager-service/src/main/java/com/crmfoodestablishment/usermanager/crud/dto/RegisterUserRequestDTO.java:
```diff
@@ -0,0 +1,47 @@
+
+import com.crmfoodestablishment.usermanager.crud.entity.Role;
+import com.fasterxml.jackson.annotation.JsonFormat;
+import jakarta.validation.constraints.*;
```
here too wildcard import

zhenyaAreYouHere63 commented on user-manager-service/src/main/java/com/crmfoodestablishment/usermanager/crud/dto/UpdateUserRequestDTO.java:
```diff
@@ -0,0 +1,46 @@
+
+import com.crmfoodestablishment.usermanager.crud.entity.Role;
+import com.fasterxml.jackson.annotation.JsonFormat;
+import jakarta.validation.constraints.*;
```
here too wildcard import. And look at all the code there are still such wildcard imports.

Commit:
Refactored auth part:
- extracted Key conversion methods into separate KeyUtils class
- fixed access token role parsing bug in AccessTokenHandlerAdapter

Commit:
Separated import statements to internal project code from external
Implemented tests:
- of auth part code
- of UserControllerAdvice

Commit:
Sync commit

Commit:
Refactored exception handling according to agreed format of api error

Commit:
Removed redundant check in controllers for nil uuid

Commit:
Renamed access and refresh tokens expiration time property to expirationTimeInMinutes

Commit:
Added AuthService dependency to UserServiceImpl to logout users when deleting

momopoi443 commented on user-manager-service/src/test/java/com/crmfoodestablishment/usermanager/crud/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,271 @@
+                .andExpect(jsonPath("$.description", notNullValue()));
+    }
+
+    private UpdateUserRequestDTO mapRegisterUserRequestDTOToUpdateUserRequestDTO(
```
Private methods always located in the bottom of the class

zhenyaBaraniuk commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+            );
+        }
+
+        if (parsedAccessToken.claims().exp().isBefore(LocalDateTime.now())) {
```
maybe isAfter()?

momopoi443 replied:
exp claim in JWT token is date to what it is considered valid, so if expiration date is after now JWT is valid. So to check if JWT is valid we need to ensure that its expiration date is not before now

zhenyaBaraniuk commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/KeyUtils.java:
```diff
@@ -0,0 +1,30 @@
+            String token,
+            KeyFactory keyFactory
+    ) throws InvalidKeySpecException {
+        byte[] decodedRefreshTokenSecretKey = Base64.getDecoder().decode(token);
```
'decodedRefreshTokenSecretKey' why is there a RefreshToken, accessToken is also there converted here.

zhenyaBaraniuk commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/KeyUtils.java:
```diff
@@ -0,0 +1,30 @@
+            String token,
+            KeyFactory keyFactory
+    ) throws InvalidKeySpecException {
+        byte[] decodedRefreshTokenPublicKey = Base64.getDecoder().decode(token);
```
here, too, with the name of the variable.

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,64 @@
+
+    @PostMapping("/logout/{userUuid}")
+    public ResponseEntity<Void> logout(
+            @PathVariable
```
pls combine it in 1 line

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/dto/CredentialsDTO.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+@Setter
+public class CredentialsDTO {
```
it would be a record

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/dto/UserDTO.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof UserDTO teacher)) return false;
```
rename this teacher

Commit:
Refactored JwtServiceImpl:
- refactored names, tabs etc.
- removed redundant check of jwt exp claim
Refactored KeyUtils

Commit:
Made CredentialsDTO a record
Refactored AuthServiceImpl:
- due to CredentialsDTO changes
- refactored Optional usage

Commit:
ApiErrorInfo:
- renamed to ApiErrorDTO
- made a record
AuthController:
- changed name of logout path variable to userId
- refactored code style
AuthControllerAdvice:
- changed due to ApiErrorDTO
- refactored code style
GlobalInputValidationControllerAdvice:
- same as with AuthControllerAdvice
- removed redundant log statements

Commit:
ApiErrorInfo:
- renamed to ApiErrorDTO
- made a record
AuthController:
- changed name of logout path variable to userId
- refactored code style
AuthControllerAdvice:
- changed due to ApiErrorDTO
- refactored code style
GlobalInputValidationControllerAdvice:
- same as with AuthControllerAdvice
- removed redundant log statements
UserControllerAdvice:
- same as with AuthControllerAdvice

Commit:
UserController:
- removed wildcard imports
- refactored code style

Commit:
Merge remote-tracking branch 'origin/UAS-Implemented-crud-user-api-and-user-auth' into UAS-CRM-19-Implemented-crud-user-api-and-user-auth

Commit:
Made all dtos records
Refactored UserMapper tabs
Changed UserMapperImpl due to dtos change
Refactored UserService method params names
UserServiceImpl:
- refactored to all above-mentioned changes
- refactored Optional usage

Commit:
JwtServiceImplTest:
- refactored test names
- refactored due to dtos record methods
- added assertions on thrown exception messages
- refactored overall code style

Commit:
Validation:
- added classes of regexp and error message constants
- accordingly changed dtos and controller method arguments
- refactored GlobalInputValidationControllerAdvice
Update and register dtos:
- merged together
- accordingly refactored userMapper, UserService and UserController
Refactored exception message in AuthServiceImpl refresh()
Made TokenPair a record

Commit:
Refactored all tests
Added tests for GlobalInputValidationAdvice

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/UserManagerApplication.java:
```diff
@@ -1,11 +1,15 @@
 @EnableDiscoveryClient
+@ConfigurationPropertiesScan
+
+//TODO add tests for logs
```
Why do you need tests for logs?

momopoi443 replied:
IMO, It will be food to add log checks in existing unit tests

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,60 @@
+@Validated
+public class AuthController {
+
+    public static final String AUTH_LOGIN_PATH = "/api/auth/login";
```
please remove this constants from controller and move them to the tests

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,60 @@
+    public ResponseEntity<TokenPair> login(@RequestBody @Valid CredentialsDTO credentials) {
+        TokenPair tokenPair = authService.login(credentials);
+
+        return ResponseEntity.ok().body(tokenPair);
```
This might be simplified: 
ResponseEntity.ok(tokenPair)

BTW, why do you need ResponseEntity here? You don't use any headers, different statuses, or anything that would justify using it

momopoi443 replied:
As for me, I expect to see status in return statement rather than in annotation above method signature

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,60 @@
+    @PostMapping("/refresh")
+    public ResponseEntity<String> refresh(
+            @RequestParam @Pattern(regexp = REFRESH_TOKEN_REGEXP, message = REFRESH_TOKEN_MESSAGE)
+            String refreshToken
```
pls move this line to the previous one (after annotations)

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/controller/AuthControllerAdvice.java:
```diff
@@ -0,0 +1,34 @@
+@RestControllerAdvice
+public class AuthControllerAdvice {
+
+    @ExceptionHandler({InvalidTokenException.class})
```
you can remove '{}'

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/dto/CredentialsDTO.java:
```diff
@@ -0,0 +1,21 @@
+import static com.crm.food.establishment.user.validation.ValidationRegexps.EMAIL_REGEXP;
+import static com.crm.food.establishment.user.validation.ValidationRegexps.PASSWORD_REGEXP;
+
+public record CredentialsDTO (
```
please change DTO to Dto

momopoi443 replied:
But this is abbreviation, I thought they need to written in upper-case

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/ApiErrorDTO.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Builder;
+
+@Builder
+public record ApiErrorDTO (
```
pls don't use upper case for DTO

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/controller/AuthControllerAdvice.java:
```diff
@@ -0,0 +1,34 @@
+        return new ResponseEntity<>(errorInfo, HttpStatus.UNAUTHORIZED);
+    }
+
+    @ExceptionHandler({InvalidUserCredentialsException.class})
```
the same is here - you can remove '{}'

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/validation/ValidationErrorMessages.java:
```diff
@@ -0,0 +1,22 @@
+
+public class ValidationErrorMessages {
+
+    public static final String EMAIL_REGEXP_MESSAGE = "Given invalid email";
```
i suggest to rename this constant to be INVALID_EMAIL_MESSAGE or smth like this

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/validation/ValidationErrorMessages.java:
```diff
@@ -0,0 +1,22 @@
+package com.crm.food.establishment.user.validation;
+
+public class ValidationErrorMessages {
```
this class might be an interface + you will be able to remove redundant 'public static final'

momopoi443 replied:
I understand but is it intuitive to have such interface?

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/validation/ValidationRegexps.java:
```diff
@@ -0,0 +1,10 @@
+
+public class ValidationRegexps {
+
+    public static final String EMAIL_REGEXP = "^[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$";
```
looks impressively. I believe it can be simplified

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/validation/ValidationRegexps.java:
```diff
@@ -0,0 +1,10 @@
+package com.crm.food.establishment.user.validation;
+
+public class ValidationRegexps {
```
this class can be an interface as well, but considering that they are used only in one place, it is redundant to move them to constants

momopoi443 replied:
Actually not:
<img width="455" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/80dbd547-7c78-419f-a031-5c93b0a09261">
<img width="288" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/7609f66c-5162-4cb9-9714-ce4c6f20a272">
And I thought It would be good practice to keeps them in one place

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/validation/ValidationRegexps.java:
```diff
@@ -0,0 +1,10 @@
+
+    public static final String EMAIL_REGEXP = "^[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$";
+
+    public static final String PASSWORD_REGEXP = "^[A-Za-z0-9]*.{8,}$";
```
why password should be started from a letter or a digit?

momopoi443 replied:
Why it should not? :)
So I think we need to come up with requirements for passwords and emails

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/dto/CredentialsDTO.java:
```diff
@@ -0,0 +1,21 @@
+        String email,
+
+        @Pattern(regexp = PASSWORD_REGEXP, message = PASSWORD_REGEXP_MESSAGE)
+        @NotBlank(message = NOT_BLANC_MESSAGE)
```
typo - should be blank

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/KeyUtils.java:
```diff
@@ -0,0 +1,30 @@
+
+public class KeyUtils {
+
+    public static PrivateKey convertBase64PrivateKeyToJavaPrivateKey(
```
imo it is better to move first param to the 13th line
![image](https://github.com/polinaucc/crm-food-establishment/assets/31660519/344b8e6b-8c55-46b1-af57-20b63d691e17)

momopoi443 replied:
I understand that you suggest it from perspective of your experience, but this looks strange for me. 'Cause as it would be in unreadable in case of writing in one line, it remains unreadable in your suggestion

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/KeyUtils.java:
```diff
@@ -0,0 +1,30 @@
+        return keyFactory.generatePrivate(pkcs8PrivateKeySpec);
+    }
+
+    public static PublicKey convertBase64PublicKeyToJavaPublicKey(
```
the same is here

momopoi443 replied:
My opinion in previous comment

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/TimeUtils.java:
```diff
@@ -0,0 +1,20 @@
+public class TimeUtils {
+
+    public static Date convertLocalDateTimeToDate(LocalDateTime dateTime) {
+        return Date.from(
```
imo it can be written in 1 line

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/AuthServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+            throw new InvalidUserCredentialsException("Wrong password");
+        }
+
+        log.info("User: " + foundUser.getEmail() + " logged");
```
logged in (?)
+ I suggest to wrap value (in this case - email) in [ ] + add info that is email. F.e. User with email [blblabbb@gmail.com] is logged in

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/AuthServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+                        "Invalid subject: no users with id + " + parsedRefreshToken.claims().sub()
+                ));
+
+        log.info("User: " + foundUser.getUuid() + " refreshed access token");
```
the same is here. I would add here info that it is 'uuid' of the user

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/AuthServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+    @Override
+    public void logout(UUID userUuid) {
+        jwtService.invalidateRefreshToken(userUuid);
+        log.info("User: " + userUuid + " logged out");
```
the same is here

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+        accessTokenParser = Jwts.parserBuilder()
+                .setSigningKey(accessTokenPublicKey)
+                .build();
+        accessTokenHandlerAdapter = new AccessTokenHandlerAdapter();
```
i would add here this as well to improve readability (in all class members that are mentioned in a constructor)

momopoi443 replied:
What you mean by 'this', logs?

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+
+    @Override
+    public AccessToken parseAccessToken(String accessToken) {
+        AccessToken parsedAccessToken;
```
this line is redundant. Try block may look like:
```
 try {
            return accessTokenParser.parse(accessToken, accessTokenHandlerAdapter);
        }
```

+ return from 162 line will be removed

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+        AccessToken parsedAccessToken;
+
+        try {
+            parsedAccessToken = accessTokenParser.parse(
```
pls write that in 1 line

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+        RefreshToken parsedRefreshToken;
+
+        try {
+            parsedRefreshToken = refreshTokenParser.parse(
```
pls write this in 1 line

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+
+        if (Boolean.FALSE.equals(refreshTokenRedisTemplate.hasKey(
+                parsedRefreshToken.claims().sub()
+        ))) {
```
pls move 181 line to the previous one

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/token/adapter/RefreshTokenHandlerAdapter.java:
```diff
@@ -0,0 +1,28 @@
+                        convertDateToLocalDateTime(jws.getBody().getIssuedAt()),
+                        convertDateToLocalDateTime(jws.getBody().getExpiration()),
+                        UUID.fromString(
+                                jws.getBody().getSubject()
```
pls move that to the 22 line

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/controller/UserController.java:
```diff
@@ -0,0 +1,77 @@
+@Validated
+public class UserController {
+
+    public static final String USER_PATH = "/api/user";
```
pls move this constants to the tests

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/controller/UserControllerAdvice.java:
```diff
@@ -0,0 +1,34 @@
+                .description(exception.getMessage())
+                .build();
+
+        return new ResponseEntity<>(errorDTO, HttpStatus.BAD_REQUEST);
```
I would move response status to the annotation above the method and return only error. Imo it will simplify the code

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/mapper/UserMapperImpl.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Override
+    public void mapUpdateRegisterUserRequestDTOToUser(
```
it is unexpected to see void type in method that should map something from 1 type to another one

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/mapper/UserMapperImpl.java:
```diff
@@ -0,0 +1,48 @@
+
+@Component
+@RequiredArgsConstructor
+public class UserMapperImpl implements UserMapper {
```
This implementation should be added by Mapstruct
To inject and use PasswordEncoder there you can add the following:

```
@Mapper(componentModel = "spring", uses = PasswordEncoder.class)
public interface UserMapper {
    
    @Mapping(target = "password", expression = "java(passwordEncoder.encode(dto.getPassword()))")
    User mapUpdateRegisterUserRequestDTOToUser(UpdateRegisterUserRequestDTO dto);
}
```

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/KeyUtils.java:
```diff
@@ -0,0 +1,30 @@
+    public static PrivateKey convertBase64PrivateKeyToJavaPrivateKey(
+            String base64PrivateKey,
+            KeyFactory keyFactory
+    ) throws InvalidKeySpecException {
```
no need to move this to new line

momopoi443 replied:
If mean below code, for me it looks highly unstructured
<img width="417" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/b053131a-3e54-47b4-8d9b-c5916ef14b68">

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+    public JwtServiceImpl(
+            JwtProperties jwtProperties,
+            RedisTemplate<UUID, String> refreshTokenRedisTemplate
+    ) throws NoSuchAlgorithmException, InvalidKeySpecException {
```
It would be better t make it 1 line

momopoi443 replied:
You mean this:
<img width="752" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/f1f187e5-53a2-4680-9999-6a64c9e0af5f">
I think it is way too long

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+
+    @Override
+    public String issueRefreshToken(User user) {
+        LocalDateTime issuedTime = LocalDateTime.now().withSecond(0).withNano(0);
```
why do you set `withSecond(0)` ?

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/auth/service/impl/JwtServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+        LocalDateTime issuedTime = LocalDateTime.now().withSecond(0).withNano(0);
+        LocalDateTime expirationTime = issuedTime.plusMinutes(
+                jwtProperties.refreshToken().expirationTimeInMinutes()
+        );
```
no need to put braces to new line here

momopoi443 replied:
I wanted to separate extraction of 'expirationTimeInMinutes', so it will be visibly highlighted. If you mean put it in one line, first of all it will be long, secondly it is hard to read for me:
<img width="702" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/3c5136ae-310a-4199-b8fe-41fbcd3d6b8f">

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/controller/UserController.java:
```diff
@@ -0,0 +1,77 @@
+
+    @PostMapping
+    public ResponseEntity<RegisterUserResponseDTO> registerUser(
+            @RequestBody @Valid UpdateRegisterUserRequestDTO creationDTO
```
Same thing about new lines, there is no need to put them here

momopoi443 replied:
IMO, it is way too long, when I need to run my eyes along the entire length of endpoint signature, it hearts. I am not joking)
<img width="766" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/302786cb-956e-4883-8b1d-30d9ecd4a943">
If you mean this, for me it look unstructured, and decrease readability: 
<img width="506" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/5f810700-14c5-42be-917d-16ee44c9128c">

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/dto/UpdateRegisterUserRequestDTO.java:
```diff
@@ -0,0 +1,57 @@
+        String lastName,
+
+        @NotNull(message = NOT_NULL_MESSAGE)
+        Boolean isMale,
```
IMO, it would be more readable if you make it Enum with 2 options instead of Boolean 
BTW maybe it make sense to remove this field from here? Because it doesn't make have any use in scope of this domen

momopoi443 replied:
1) About enum I think it would be redundant here, it would rather make it raw type boolean
2) About existence of this field, I think wee need to discuss It all together, 'cause I took it from ERD of Andrii

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/entity/User.java:
```diff
@@ -0,0 +1,54 @@
+package com.crm.food.establishment.user.manager.entity;
+
+import jakarta.persistence.*;
```
Please remove `.*`

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Override
+    public void update(UUID userUuid, UpdateRegisterUserRequestDTO updateDTO) {
+        AtomicReference<User> userToUpdate = new AtomicReference<>();
```
To use atomic reference just to use it in lambda is not a good practice imo. It would be better to rewrite that piece of code without lambda.

momopoi443 replied:
First, can you explain why usage of atomic reference is bad practice. Second, I not trying just to use lambda, I want to check if owner trying to update its record. 
I can rewrite it like this, you think it is better?
<img width="480" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/4e11d54b-9cdc-47ca-a958-3a162cd199ec">

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Override
+    public void update(UUID userUuid, UpdateRegisterUserRequestDTO updateDTO) {
```
my preference is to add info about what to save 
we discussed it today. F.e updateUser

momopoi443 replied:
I thought It would be transparently that create method in UserService creates users. But ok, I can change it

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+            userToDelete.setRole(Role.CLIENT);
+            userRepository.save(userToDelete);
+
+            log.info("Fired " + userToDeleteRole.name() + ": " + userUuid);
```
my suggestions is to change this log to smth like: "Role for user [user-id] was changed from [role-old] to [role-new] "

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+            userRepository.delete(userToDelete);
+            authService.logout(userToDelete.getUuid());
+
+            log.info("Deleted user: " + userUuid);
```
it is better to add log messages in format: User [user-id] was successfully deleted 
what + details + what was done?

momopoi443 replied:
By " + details + what was done" you mean to say that it also was logged out? I think it would be redundant 'cause AuthService methods also has logs

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Override
+    public List<UserDTO> listAll() {
```
rename it to something like `getAll`

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+        Optional<User> user = userRepository.findByEmail(creationData.email());
+        User userToRegister;
+
+        if (
```
same thing with braces here

momopoi443 replied:
It would be too long to write it in one line, and in other ways it would be less readable in my opinion

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/manager/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+                .toList();
+    }
+
+    private User createUser(UpdateRegisterUserRequestDTO creationData) {
```
If I understand this correctly this method is creating new user and updating it in some cases 
It should be 2 separate methods
Also please consider that if someone else register user, he doesn't create password

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/validation/GlobalInputValidationControllerAdvice.java:
```diff
@@ -0,0 +1,39 @@
+    public ResponseEntity<List<ApiErrorDTO>> handleMethodArgumentNotValid(MethodArgumentNotValidException exception) {
+        List<ApiErrorDTO> errorInfos = exception.getFieldErrors().stream()
+                .map(error -> new ApiErrorDTO(
+                        InvalidArgumentException.errorCode(),
```
This exception can potentially be thrown in different cases and in those cases error code might be different, so IMO it would be better to pass error code from the place where we throw this exception

momopoi443 replied:
Problem is that I don't have access to code that throws this exception, code is library. And as you can see in docs of this exception, it is thrown when @Valid fails:
<img width="278" alt="image" src="https://github.com/polinaucc/crm-food-establishment/assets/95944057/db703cfc-fa82-4ec0-bc75-4e6e3a577287">

polinaucc commented on user-manager-service/src/main/java/com/crm/food/establishment/user/validation/GlobalInputValidationControllerAdvice.java:
```diff
@@ -0,0 +1,39 @@
+    @ExceptionHandler(ConstraintViolationException.class)
+    public ResponseEntity<ApiErrorDTO> handleConstraintViolationException(ConstraintViolationException exception) {
+        ApiErrorDTO errorInfo = ApiErrorDTO.builder()
+                .code(InvalidArgumentException.errorCode())
```
same thing here

momopoi443 replied:
Same problem it is thrown by library. Also I think semantically my error code suits this exception

polinaucc commented on user-manager-service/src/main/resources/application.properties:
```diff
@@ -1,10 +1,24 @@
-spring.datasource.password=${USERMANAGER_DATABASE_PASSWORD}
-spring.liquibase.change-log=classpath:db/changelog/db.changelog-master.xml
-spring.jpa.hibernate.ddl-auto=validate
+spring.datasource.username=${USER_AUTH_SERVICE_POSTGRES_USERNAME}
```
It would be nice to have application-dev.properties so you can simply run this application and will not need to crete all this env variables

polinaucc commented on user-manager-service/src/main/resources/application.properties:
```diff
@@ -1,10 +1,24 @@
-spring.application.name=user-manager-service
+spring.application.name=user-auth-service
```
why do you use .properties file instead of yaml?

momopoi443 replied:
My preference, I more like brackets than python tabs

polinaucc commented on user-manager-service/src/test/java/com/crm/food/establishment/user/auth/controller/AuthControllerAdviceTest.java:
```diff
@@ -0,0 +1,96 @@
+                ApiErrorDTO.class
+        );
+
+        response.andExpect(status().isUnauthorized())
```
you can put all those .andExpect directly after you get response

momopoi443 replied:
I like to have separate section of assertions, and not combine it all together

polinaucc commented on user-manager-service/src/test/java/com/crm/food/establishment/user/manager/controller/UserControllerAdviceTest.java:
```diff
@@ -0,0 +1,78 @@
+
+        response.andExpect(status().isBadRequest())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON));
+        assertEquals(expectedErrorDTO, actualErrorDTO);
```
FYI you can lookup json path here, sometimes it's more convenient than creating a full entity

momopoi443 replied:
I know, and in old version I used json path, but I it looks better

Commit:
Resolved comments on pull request
